### PR TITLE
Conformance tests for BLS12_381 multiScalarMul operations (2)

### PR DESCRIPTION
**Pull request by kwxm**
Conformance tests for `bls12_381_G1_multiScalarMul` and `bls12_381_G2_multiScalarMul` (20 tests each).  Each test case has a comment saying what it does.

Also fixes an incorrect size measure for lists in `plutus-metatheory`.

This is a replacement for #7366 to try to overcome some CI problems.

Closes https://github.com/IntersectMBO/plutus/issues/7325.
Closes https://github.com/IntersectMBO/plutus-private/issues/1440.

### CODE REVIEW

File: plutus-metatheory/src/Cost/Size.lagda.md [hunk 1]
```diff
@@ -74,7 +74,6 @@ defaultConstantMeasure (tmCon (atomic aData) d) = dataSize d
 defaultConstantMeasure (tmCon (atomic aBls12-381-g1-element) x) = g1ElementSize x
 defaultConstantMeasure (tmCon (atomic aBls12-381-g2-element) x) = g2ElementSize x
 defaultConstantMeasure (tmCon (atomic aBls12-381-mlresult) x) = mlResultElementSize x
-defaultConstantMeasure (tmCon (list t) []) = 1
```
Comment (kwxm) [inline plutus-metatheory/src/Cost/Size.lagda.md:77]: This was incorrect.  See [this comment](https://github.com/IntersectMBO/plutus/blob/2544c0c953f41d6827f645cd1733a33572184c6d/plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs#L207).  The size measure for the empty list was earlier changed from 0 to 1 in #6368; I think this was because we previously had a size measure for lists that included the sizes of the elements, but changed it to just be the length in #7049 and forgot to fix the Agda version.

File: plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/bls12_381_G1_multiScalarMul/multiScalarMul-01/multiScalarMul-01.uplc [hunk 2]
```diff
@@ -0,0 +1,13 @@
+-- Multi-scalar multiplication is the same as ordinary scalar multiplication for inputs of size 1.
+(program 1.0.0
+ [(builtin bls12_381_G1_equal)
+  [(builtin bls12_381_G1_multiScalarMul)
+   (con (list integer) [-7843724524521392138901923801823923123123454352157])
+   (con (list bls12_381_G1_element) [0x950dfd33da2682260c76038dfb8bad6e84ae9d599a3c151815945ac1e6ef6b1027cd917f3907479d20d636ce437a41f5])
```
Comment (ana-pantilie) [inline plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/bls12_381_G1_multiScalarMul/multiScalarMul-01/multiScalarMul-01.uplc:6]: Do these inputs mean anything or are they just random?

File: plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/bls12_381_G1_multiScalarMul/multiScalarMul-02b/multiScalarMul-02b.uplc [hunk 3]
```diff
@@ -0,0 +1,61 @@
+-- Multi-scalar multiplication is itertated multiply-and-add.
+-- msm [n1,n2,n3,n4,n5,n6] [p1,p2,p3,p4,p5,p6] = n1*p1+(n2*p2+(n3*p3+(n4*p4+(n5*p5+n6*p6))))
+-- No infinite points in the input.
```
Comment (ana-pantilie) [inline plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/bls12_381_G1_multiScalarMul/multiScalarMul-02b/multiScalarMul-02b.uplc:3]: `0xc00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000` is an infinite point?

File: plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/bls12_381_G1_multiScalarMul/multiScalarMul-03a/multiScalarMul-03a.uplc [hunk 4]
```diff
@@ -0,0 +1,60 @@
+-- Multi-scalar multiplication is itertated multiply-and-add.
+-- msm [n1,n2,n3,n4,n5,n6] [p1,p2,p3,p4,p5,p6] = ((((n1*p1+n2*p2)+n3*p3)+n4*p4)+n5*p5)+n6*p6
```
Comment (ana-pantilie) [inline plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/bls12_381_G1_multiScalarMul/multiScalarMul-03a/multiScalarMul-03a.uplc:2]: If you know `bls12_381_G1_add` is associative I don't think you need both the associated to the left and the associated to the right test.

File: plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/bls12_381_G1_multiScalarMul/multiScalarMul-04a/multiScalarMul-04a.uplc [hunk 5]
```diff
@@ -0,0 +1,60 @@
+-- Multi-scalar multiplication is itertated multiply-and-add, invariant under permutation and association.
+-- msm [n1,n2,n3,n4,n5,n6] [p1,p2,p3,p4,p5,p6] = n4*p4+((n3*p3+n1*p1)+(n6*p6+(n2*p2+n5*p5)))
```
Comment (ana-pantilie) [inline plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/bls12_381_G1_multiScalarMul/multiScalarMul-04a/multiScalarMul-04a.uplc:2]: Similar to the above, if you know `bls12_381_G1_add` is commutative and associative then the test can be subsumed by one of the above.

File: plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/bls12_381_G1_multiScalarMul/multiScalarMul-06a/multiScalarMul-06a.uplc [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    [ 0xab5d1d67b495361c3297c721cf3c9dc510fc5055bdf92fefc1e67d91a00a765520150428eb4c2fb01902d41d5ca62d85
+    , 0xc00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
+    , 0xb1cdc04561c6bd93294fc05a3a27170ffdca767965cdd7f2ba6065d6805e9c3afbb64e68604ddce26401566a74b35fb4
+    , 0x8b9120083e12fe72d41e1488078a84a6629c2d90e00c238152b4cba64c20994196a1f4de9132f44597b566c80f17d64a
+    , 0x84a8c4b64514db1bedfab0baa53fbed998672626b4c715652583a504df9c44ad3fe106cb421694cc86b2f2371eacd28a
+    , 0x8b9120083e12fe72d41e1488078a84a6629c2d90e00c238152b4cba64c20994196a1f4de9132f44597b566c80f17d64a
+    ]
+   )
+  ]
+  (con bls12_381_G1_element 0xc00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000)
```
Comment (ana-pantilie) [inline plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/bls12_381_G1_multiScalarMul/multiScalarMul-06a/multiScalarMul-06a.uplc:17]: I don't know anything about elliptic curves, so this point represents 0? It's also an infinite point?

**Review by ana-pantilie — APPROVED** [review]
The tests look good! Approving since I don't want to block this, but I do have some general comments.

Many of the tests check properties of multiScalarMul which can be generalized to property tests. If we don't have the necessary infrastructure to add such property tests to builtins then that's fine, but maybe we should consider adding it.

Also, some of the tests aren't really checking `multiScalarMul`, they're checking the properties of `...add`, which I argue should be defined for `...add` and assumed true here.

Final comment: unit tests are good for checking corner cases you aren't sure you'll hit with property tests. Are you sure you've covered all such cases?

File: plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/bls12_381_G1_multiScalarMul/multiScalarMul-02b/multiScalarMul-02b.uplc (hunk 3 above)
Comment (kwxm) [reply plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/bls12_381_G1_multiScalarMul/multiScalarMul-02b/multiScalarMul-02b.uplc:3]: Yes.

File: plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/bls12_381_G1_multiScalarMul/multiScalarMul-06a/multiScalarMul-06a.uplc (hunk 6 above)
Comment (kwxm) [reply plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/bls12_381_G1_multiScalarMul/multiScalarMul-06a/multiScalarMul-06a.uplc:17]: Sorry.  There's a zero point 0 with the property that `add 0 g = g` for all `g`, and that's also called "the point at infinity"; its internal representation can differ from general points, so it's worth checking that it doesn't cause any problems.

File: plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/bls12_381_G1_multiScalarMul/multiScalarMul-03a/multiScalarMul-03a.uplc (hunk 4 above)
Comment (kwxm) [reply plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/bls12_381_G1_multiScalarMul/multiScalarMul-03a/multiScalarMul-03a.uplc:2]: > If you know `bls12_381_G1_add` is associative I don't think you need both the associated to the left and the associated to the right test.

Yeah, I was just trying to be thorough.  I always tend to think that it's better to have more tests even if there's some redundancy.  Also, some people may be using these for other tools that they can't use the property tests with, so maybe there's an argument for allowing a bit of overlap since it may increase coverage a bit.

File: plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/bls12_381_G1_multiScalarMul/multiScalarMul-01/multiScalarMul-01.uplc (hunk 2 above)
Comment (kwxm) [reply plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/bls12_381_G1_multiScalarMul/multiScalarMul-01/multiScalarMul-01.uplc:6]: They're random points obtained by hashing random bytestrings onto the curve using the `bls12_381G1/2_hashToCurve` functions, which convert bytestrings into curve points in a way that's thought to be intractable to reverse.

**Comment by kwxm** [comment]
> Many of the tests check properties of `multiScalarMul` which can be generalized to property tests. If we don't have the necessary infrastructure to add such property tests to builtins then that's fine, but maybe we should consider adding it as a separate endeavor.

That's been in the backlog for several years, but we've never got round to it (and we're not sure how to do it).  We do have extensive property tests [here](https://github.com/IntersectMBO/plutus/blob/master/plutus-core/untyped-plutus-core/testlib/Evaluation/Builtins/BLS12_381.hs), and added tests for the MSM functions in #7343.  Ideally some day we'll convert all of the property tests into conformance tests.

> Final comment: unit tests are good for checking corner cases you aren't sure you'll hit with property tests. Are you sure you've covered all such cases?

I hope so, but it's hard to be sure.

File: plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/bls12_381_G1_multiScalarMul/multiScalarMul-06a/multiScalarMul-06a.uplc (hunk 6 above)
Comment (ana-pantilie) [reply plutus-conformance/test-cases/uplc/evaluation/builtin/semantics/bls12_381_G1_multiScalarMul/multiScalarMul-06a/multiScalarMul-06a.uplc:17]: Thanks for explaining!

**Comment by kwxm** [comment]
I'm having trouble getting this through CI, so I've rebased to see if that helps.
